src/read/macinfo.rs

```
@@ -0,0 +1,290 @@
+    ///
+    /// It is the caller's responsibility to read the `.debug_macinfo` section and
+    /// present it as a `&[u8]` slice. That means using some ELF loader on
+    /// Linux, a Mach-O loader on macOS, etc.
+    ///
+    /// ```
+    /// use gimli::{DebugMacInfo, LittleEndian};
+    ///
+    /// # let buf = [1, 0, 95, 95, 83, 84, 68, 67, 95, 95, 32, 49, 0];
+    /// # let read_section_somehow = || &buf;
+    /// let debug_str = DebugMacInfo::new(read_section_somehow(), LittleEndian);
+    /// ```
+    pub fn new(macinfo_section: &'input [u8], endian: Endian) -> Self {
+        Self::from(EndianSlice::new(macinfo_section, endian))
+    }
+}
+
+impl<R: Reader> DebugMacInfo<R> {
+    /// Look up a macro reference the `.debug_macinfo` section by DebugMacinfoOffset.
+    ///
+    /// A macinfo offset points to a list of macro information entries in the `.debug_macinfo` section.
+    /// To handle this, the function returns an iterator.
+    ///
+    /// ```
+    /// use gimli::{DebugMacInfo, DebugMacinfoOffset, LittleEndian};
+    ///
+    /// # let buf = [1, 0, 95, 95, 83, 84, 68, 67, 95, 95, 32, 49, 0];
+    /// # let offset = DebugMacinfoOffset(0);
+    /// # let read_section_somehow = || &buf;
+    /// # let debug_macinfo_offset_somehow = || offset;
+    /// let debug_macinfo = DebugMacInfo::new(read_section_somehow(), LittleEndian);
+    /// for macinfo in debug_macinfo.get_macinfo(debug_macinfo_offset_somehow()) {
+    ///     println!("Found macro info {:?}", macinfo);
+    /// }
+    /// ```
+    pub fn get_macinfo(
+        &self,
+        offset: DebugMacinfoOffset<R::Offset>,
+    ) -> impl Iterator<Item = DebugMacInfoItem<R>> {
+        let mut input = self.section.clone();
+        if let Ok(_) = input.skip(offset.0) {
+            DebugMacInfoIterator {
+                input_reader: input,
+                finished: false,
+            }
+        } else {
+            // If the offset is invalid, the iterator is immediately finished.
```

Return the error instead of an empty iterator.